tethys_apps/base/app_base.py

```diff
@@ -131,6 +136,82 @@ def url_patterns(self):
+                        handler_patterns['websocket'][namespace] = []
+
+                    # Create django url object
+                    if isinstance(url_map.handler, str):
+                        root_handler_path = 'tethysext' if is_extension else 'tethysapp'
+                        full_handler_path = '.'.join([root_handler_path, url_map.handler])
+                        handler_parts = full_handler_path.split('.')
+                        module_name = '.'.join(handler_parts[:-1])
+                        function_name = handler_parts[-1]
+                        try:
+                            module = __import__(module_name, fromlist=[function_name])
+                        except Exception as e:
+                            error_msg = 'The following error occurred while trying to import the handler function ' \
+                                        '"{0}":\n {1}'.format(url_map.handler, traceback.format_exc(2))
+                            tethys_log.error(error_msg)
+                            raise e
+                        try:
+                            handler_function = getattr(module, function_name)
+                        except AttributeError as e:
+                            error_msg = 'The following error occurred while trying to access the handler function ' \
+                                        '"{0}":\n {1}'.format(url_map.handler, traceback.format_exc(2))
+                            tethys_log.error(error_msg)
+                            raise e
+                    else:
+                        handler_function = url_map.handler
+
+                    if url_map.handler_type == 'bokeh':
+                        if url_map.url in [r'', r'/', r'^$', r'^/$']:
+                            app_endpoint = '/'.join(['apps', self.root_url])
+                        else:
+                            stripped_url = url_map.url.replace("^", "").replace("$", "")
+                            if stripped_url.endswith('/'):
+                                stripped_url = stripped_url[:-1]
+
+                            app_endpoint = '/'.join(['apps', self.root_url, stripped_url])
+
+                        bokeh_app = autoload(app_endpoint, handler_function)
+                        kwargs = dict(app_context=bokeh_app.app_context)
+
+                        def urlpattern(suffix=""):
+                            url_pattern = bokeh_app.url + suffix
+                            return f'^{url_pattern}$'
+
+                        http_url = url(urlpattern('/autoload.js'), BokehAutoloadJsCDN,
+                                       name=f'{url_map.name}_bokeh_autoload', kwargs=kwargs)
+                        ws_url = url(urlpattern('/ws'), WSConsumer, name=f'{url_map.name}_bokeh_ws', kwargs=kwargs)
+
+                        # Append to namespace list
+                        handler_patterns['http'][namespace].append(http_url)
+                        handler_patterns['websocket'][namespace].append(ws_url)
```

Comment:
I think we'd be better served if lines187-209 were extracted into a utility function/method as well.